### ISSUE TITLE
data parallel contrastive learning

### ISSUE THREAD

**Issue by krajsek** [helmholtz-analytics/heat#1123]
**Feature functionality**
Heat contains data parallelism, which is particularly useful for loss functions that are additive separable but  not suitable for non-separable loss functions, such as those used in contrastive learning as described in https://arxiv.org/abs/2002.05709. This feature supports any type of loss function.

**Comment by mrfh92** [comment]
@krajsek how is this related to #482? - nonempty intersection, #1123 subset of #482, empty intersection?

**Comment by krajsek** [comment]
@mrfh92 This is a high level application of the AD mechanism, so it builds upon #482.  So far I have a stand alone version for PyTorch (I have send you an link to the repository in case you are interested) build on own AD MPI operations implemented in PyTorch but it should also work with the planned Heat AD as long as a AD able Send/Recv and allgather/gather OP  is available.
